core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -135,207 +146,285 @@ class PartitionStateMachine(controller: KafkaController, stateChangeLogger: Stat
-    val liveAssignedReplicas = replicaAssignment.filter(r => controllerContext.isReplicaOnline(r, topicAndPartition))
-    val stateChangeLog = stateChangeLogger.withControllerEpoch(controller.epoch)
-    liveAssignedReplicas.headOption match {
-      case None =>
-        val failMsg = s"Encountered error during state change of partition $topicAndPartition from New to Online, " +
-          s"assigned replicas are [${replicaAssignment.mkString(",")}], live brokers are " +
-          s"[${controllerContext.liveBrokerIds}]. No assigned replica is alive."
-        stateChangeLog.error(failMsg)
-        throw new StateChangeFailedException(stateChangeLog.messageWithPrefix(failMsg))
-
-      // leader is the first replica in the list of assigned replicas
-      case Some(leader) =>
-        debug(s"Live assigned replicas for partition $topicAndPartition are: [$liveAssignedReplicas]")
-        val leaderAndIsr = LeaderAndIsr(leader, liveAssignedReplicas)
-        val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controller.epoch)
-        debug(s"Initializing leader and isr for partition $topicAndPartition to $leaderIsrAndControllerEpoch")
-
-        try {
-          zkUtils.createPersistentPath(
-            getTopicPartitionLeaderAndIsrPath(topicAndPartition.topic, topicAndPartition.partition),
-            zkUtils.leaderAndIsrZkData(leaderAndIsr, controller.epoch)
-          )
-          // NOTE: the above write can fail only if the current controller lost its zk session and the new controller
-          // took over and initialized this partition. This can happen if the current controller went into a long
-          // GC pause
-          controllerContext.partitionLeadershipInfo.put(topicAndPartition, leaderIsrAndControllerEpoch)
-          brokerRequestBatch.addLeaderAndIsrRequestForBrokers(
-            liveAssignedReplicas,
-            topicAndPartition.topic,
-            topicAndPartition.partition,
-            leaderIsrAndControllerEpoch,
-            replicaAssignment,
-            isNew = true
-          )
-        } catch {
-          case _: ZkNodeExistsException =>
-            // read the controller epoch
-            val leaderIsrAndEpoch = ReplicationUtils.getLeaderIsrAndEpochForPartition(zkUtils, topicAndPartition.topic,
-              topicAndPartition.partition).get
-
-            val failMsg = s"Encountered error while changing partition $topicAndPartition's state from New to Online " +
-              s"since LeaderAndIsr path already exists with value ${leaderIsrAndEpoch.leaderAndIsr} and controller " +
-              s"epoch ${leaderIsrAndEpoch.controllerEpoch}"
-            stateChangeLog.error(failMsg)
-            throw new StateChangeFailedException(stateChangeLog.messageWithPrefix(failMsg))
+  private def doElectLeaderForPartitions(partitions: Seq[TopicAndPartition], partitionLeaderElectionStrategy: PartitionLeaderElectionStrategy):
```

Comment:
The class is a bit harder to read now with the new ZK wrapper. To make that a bit easier, could we add a comment to describe the return type and this method does? It may be worth doing that on a few other methods in this class as well.